server/sonar-server/src/main/java/org/sonar/server/ws/ws/ListAction.java

```diff
@@ -17,133 +17,80 @@
-          String controllerKey = request.mandatoryParam("controller");
-          Controller controller = context.controller(controllerKey);
-          if (controller == null) {
-            throw new IllegalArgumentException("Controller does not exist: " + controllerKey);
-          }
-          String actionKey = request.mandatoryParam("action");
-          Action action = controller.action(actionKey);
-          if (action == null) {
-            throw new IllegalArgumentException("Action does not exist: " + actionKey);
-          }
-          handleResponseExample(action, response);
-        }
-      });
-    action.createParam("controller")
-      .setRequired(true)
-      .setDescription("Controller of the web service")
-      .setExampleValue("api/issues");
-    action.createParam("action")
-      .setRequired(true)
-      .setDescription("Action of the web service")
-      .setExampleValue("search");
-  }
-
-  private static void handleResponseExample(Action action, Response response) throws IOException {
-    if (action.responseExample() != null) {
-      response
-        .newJsonWriter()
-        .beginObject()
-        .prop("format", action.responseExampleFormat())
-        .prop("example", action.responseExampleAsString())
-        .endObject()
-        .close();
-    } else {
-      response.noContent();
-    }
-  }
+  @Override
+  public void handle(Request request, Response response) throws Exception {
+    checkState(context != null && !context.controllers().isEmpty(), "Web service controllers must be loaded before calling the action");
 
-  void handleList(List<Controller> controllers, Request request, Response response) {
     boolean includeInternals = request.mandatoryParamAsBoolean("include_internals");
     JsonWriter writer = response.newJsonWriter();
     writer.beginObject();
     writer.name("webServices").beginArray();
 
     // sort controllers by path
-    Ordering<Controller> ordering = Ordering.natural().onResultOf(Controller::path);
-    for (Controller controller : ordering.sortedCopy(controllers)) {
+    Ordering<WebService.Controller> ordering = Ordering.natural().onResultOf(WebService.Controller::path);
+    for (WebService.Controller controller : ordering.sortedCopy(context.controllers())) {
       writeController(writer, controller, includeInternals);
     }
     writer.endArray();
     writer.endObject();
     writer.close();
   }
 
-  private static void writeController(JsonWriter writer, Controller controller, boolean includeInternals) {
+  @Override
+  public void setContext(WebService.Context context) {
+    this.context = context;
+  }
+
+  private static void writeController(JsonWriter writer, WebService.Controller controller, boolean includeInternals) {
     if (includeInternals || !controller.isInternal()) {
       writer.beginObject();
       writer.prop("path", controller.path());
       writer.prop("since", controller.since());
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png 'Severity: CRITICAL') Define a constant instead of duplicating this literal "since" 3 times. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1192)